sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -490,15 +491,18 @@ export class BatchingReceiverLite {
       reject(err);
     }, args.abortSignal);
 
-    logger.verbose(
-      `${loggingPrefix} Adding credit for receiving ${args.maxMessageCount} messages.`
-    );
+    const emptySlots = numberOfEmptyIncomingSlots(receiver);
+    const creditsToAdd =
+      this._receiveMode === "peekLock"
+        ? Math.min(args.maxMessageCount, emptySlots <= 1 ? 0 : emptySlots - 1)
+        : args.maxMessageCount;
+    logger.verbose(`${loggingPrefix} Adding credit for receiving ${creditsToAdd} messages.`);
 
-    // By adding credit here, we let the service know that at max we can handle `maxMessageCount`
+    // By adding credit here, we let the service know that at max we can handle `creditsToAdd`
     // number of messages concurrently. We will return the user an array of messages that can
-    // be of size upto maxMessageCount. Then the user needs to accordingly dispose
+    // be of size upto `creditsToAdd`. Then the user needs to accordingly dispose
     // (complete/abandon/defer/deadletter) the messages from the array.
-    receiver.addCredit(args.maxMessageCount);
+    receiver.addCredit(creditsToAdd);
```

Comment:
When we add 0 credits, what is the user experience? Is the promise returned by `receiveMessages()` resolved immediately or the user gets a timeout error?

Reply:
Currently, it'd wait for the `maxWaitTime` before returning zero messages.
Returning immediately makes more sense, I'll look into it.

Reply:
> Returning immediately makes more sense, I'll look into it.

Do you mean for the promise to be resolved or rejected?

I would recommend rejecting with an error that is retryable so that by the time our retry attempts are done with the delays in between, we give the user a chance to settle a few messages

Reply:
> Do you mean for the promise to be resolved or rejected?

I meant resolved.

> I would recommend rejecting with an error that is retryable 

I see. That sounds good too, will work on that.

Reply:
Are you thinking of treating it like a throttling error, essentially? The user does have to intervene here - should we give them a type or `.name` for this error so they can do something?

Reply:
Shouldn't it just be `GeneralError`? 
If we want to add a new name for this case - how about `BufferLimitReached`?

Reply:
Added "ExcessUnsettledMessagesInBuffer".

Reply:
Is this fine?

Reply:
Resolving here.
Since more discussions happening at https://github.com/Azure/azure-sdk-for-js/pull/14060#discussion_r620643333